### PR TITLE
Heisenbridge

### PR DESCRIPTION
**Pull request by mgoerens**

### CODE REVIEW

File: controllers/synapse/synapse_configmap.go [hunk 1]
```diff
@@ -1122,7 +1122,7 @@ url_preview_accept_language:
 
 # Enable registration for new users.
 #
-#enable_registration: false
+enable_registration: true
```
Comment (acmenezes) [inline controllers/synapse/synapse_configmap.go:1125]: This change seems unrelated to the rest of the code. What is it for? Could be done in a separate commit with some comments to clarify what is happening here.

**Review by acmenezes — COMMENTED** [review]
In general the code seems good. I made some suggestions and asked some questions. In the end it's up to you if you want to keep it like this for now or refactor a bit to make it easier to understand. I think the controller file is too big and can be separated in multiple files. Another suggestion is that you add good commit messages along side with smaller and more commits for each specific change. That helps reviewing. Adding comments to the functions and what they are doing also helps getting the idea faster. The PR size for a new feature like this is ok though. We can try a few smaller PRs next time if it is the case.

File: controllers/synapse/synapse_configmap.go (hunk 1 above)
Comment (mgoerens) [reply controllers/synapse/synapse_configmap.go:1125]: I moved it to a separate commit

File: controllers/synapse/synapse_controller.go [hunk 2]
```diff
@@ -441,7 +542,7 @@ func (r *SynapseReconciler) updateSynapseStatusDatabase(
 	return nil
 }
 
-func (r *SynapseReconciler) updateSynapseConfigMap(
+func (r *SynapseReconciler) updateSynapseConfigMapWithPostgreSQLInfos(
 	ctx context.Context,
 	cm *corev1.ConfigMap,
```
[the remark this answers is not in the document]
Comment (mgoerens) [reply controllers/synapse/synapse_controller.go]: So, I reworked the ConfigMap functions to reduce the code duplication (and therefore the length of the controller.go file - I agree it's too long and not specific enough). I used a similar model as for the `reconcile.go` file.

I left this change in a fixup commit (I think it's easier for the reviewer), but let me know if you prefer me to squash it directly.

File: controllers/synapse/synapse_controller.go [hunk 3]
```diff
@@ -523,9 +646,119 @@ func (r *SynapseReconciler) updateSynapseConfigMapData(
...
+		return nil, err
+	}
+	if err := yaml.Unmarshal(intermediate, &out); err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+func (r *SynapseReconciler) updateHeisenbridgeConfigMapURL(
```
[the remark this answers is not in the document]
Comment (mgoerens) [reply controllers/synapse/synapse_controller.go]: I'm not sure if we will have a lot of function common to all (or even some) bridges. It's hard to tell for now really. The deployment method are quite various.

I have however moved all functions related to the homeserver ConfigMap into `synapse_configmap.go` and the ones related to the heisenbridge ConfigMap into `heisenbridge_configmap.go`.

I was also thinking of moving the *files* specific to the Heisenbridge to a dedicated *directory* (`synapse_heisenbridge_configmap.go`, `synapse_heisenbridge_deployment.go`, and `synapse_heisenbridge_service.go`).

**Review by acmenezes — APPROVED** [review]
/lgtm

**Pull request merged**